library/src/scala/NamedTuple.scala

```diff
@@ -0,0 +1,24 @@
+package scala
+import annotation.experimental
+
+@experimental
+object NamedTuple:
+
+  opaque type Element[name <: String & Singleton, A] >: A = A
+
+  object Element:
+    def apply[S <: String & Singleton, A](name: S, x: A): Element[name.type, A] = x
+    def extract[S <: String & Singleton]: ValueExtractor[S] = ValueExtractor[S]()
+    extension [S <: String & Singleton, A](named: Element[S, A]) def value: A = named
+
+    class ValueExtractor[S <: String & Singleton]:
+      def unapply[A](x: Element[S, A]): Some[A] = Some(x)
```

Comment:
why does this need to be in `ValueExtractor` and not `Element.unapply`?

Reply:
I think because they're not symmetrical: you're only extracting the value here, but Element apply applies both the name and the value.

Reply:
Type inference, we need to curry the type parameters of the `unapply` because we know the first one (the name) at desiugaring but not the second one, so the second one needs to be inferred.